### ISSUE TITLE
Support of additional point dimensions (e.g. classification) in output Cesium model.

### ISSUE THREAD

**Issue by devdylan** [connormanning/entwine#118]
Hello again,

Sorry if this is addressed somewhere, but it's not clear to me if the current version of Entwine supports the preservation of non-color, non-position point dimensions (e.g. classification) when converting to Cesium models. 

I assume this information would live in the [batch table](https://github.com/AnalyticalGraphicsInc/3d-tiles/blob/master/TileFormats/BatchTable/README.md) of the output cesium 3d tiles.

It's hinted at in the discussion of the following issues & PRs, but a clear answer is not provided, nor could I find it in the documentation:
- https://github.com/connormanning/entwine/issues/59
- https://github.com/connormanning/entwine/pull/85
- https://github.com/connormanning/entwine/issues/81

Thanks for your help.

**Comment by connormanning** [comment]
This was implemented in #85 as you mentioned, but I haven't much knowledge in using the batch table feature of cesium so I didn't port it over to the new format when I did #98 because I didn't readily have a means of testing it.  However this functionality is in the Entwine 1.3 release, and its documentation is [here](https://github.com/connormanning/entwine/blob/1.3.0/entwine/formats/cesium/README.md#batch-tables), so you can use the 1.3 release for your use-case in the meantime.  It would probably be quite easy to port the functionality to the new version since entwine's cesium code is much simpler now - it just hasn't been done yet.

**Comment by devdylan** [comment]
Thanks again for the prompt reply and for pointing me toward the documentation - I didn't see that before. I'll give it a shot with Entwine 1.3.

**Comment by devdylan** [comment]
I assume this is part of the `convert` step, not the `build step` - during the `build` step, I simply add the dimension (which I eventually want to appear in the batch table) to the new `schema`.

Assuming the above, my issue is that I am not able to pass a config file to the `convert` step.   I assume this is the same "cesium format configuration" file referenced in the doc you provided. I referenced this issue in a previous issue: https://github.com/connormanning/entwine/issues/117

I've tried `-c`: 
`Encountered an error: Invalid argument: --config`
And `--config`:
`Encountered an error: Invalid argument: -c`

I can pass a config to the `build` step using `-c` without issue.

**Comment by connormanning** [comment]
I think there is a version mixup happening here - I'll try to clarify:

### Entwine 1.3
At this time, 3D Tiles output was built during the `build` phase as a separate output from entwine.  A sample workflow for this version of entwine was [here](https://github.com/connormanning/entwine-cesium-pages/tree/b825435e2da1abe292351af73f0a3e9af4d52e40).

The 3D Tiles output was triggered during `build` via a configuration like [this](https://github.com/connormanning/entwine/blob/1.3.0/config/cesium.json).  In this version, you would run `entwine build <some-cesium-config.json> -i ... -o ...`, using a configuration file as the first argument to `entwine build` (i.e. *not* with the `-c` or `--config` flag).  `entwine convert` did not exist at this time since the 3D Tiles were generated during `build`.

This version supported batch table entries via the docs I posted earlier - using a modified cesium config during `build`.

### Entwine latest
Now, an EPT dataset built via `entwine build` may be converted to 3D Tiles as a separate step as long as the build was reprojected to `EPSG:4978`.  To do so, you would run `entwine convert -i <path-to-built-EPT> ...`.  This version does *not* currently support batch table entries.

Regarding your most recent post, I see now that I haven't properly hooked up the `-c`/`--config` flag to the other applications aside from `build`, so I will fix that now.  However since this current version doesn't support batch tables, you will need to follow the 1.3 workflow regardless if you need batch tables.

**Comment by devdylan** [comment]
Ah, I was indeed mixed up between the versions. I'll pull 1.3 and follow the 1.3 directions. Much appreciated.

**Comment by devdylan** [comment]
Alright, so using the 1.3 Docker image (question on this below), I can specify the `batchTable` array in the cesium format, and it appears that the dimension is being written to the output Cesium batch table. The only way I am aware of verifying the output Cesium batch table is by styling, e.g.
```
 tileset.style = new Cesium.Cesium3DTileStyle({
        color: {
            conditions: [
                ["${Classification} >= 0", "rgb(255, 82, 198)"],
                ["${Classification} >= 1", "rgb(0, 0, 0)"],
                ["true", "rgb(255, 255, 255)"]
            ]
        }
    });
```

However, from experimenting with the styles, it looks like Classification is set to `0` across the board. Cesium throws an error if I try with `${BadDimension}`, so I have confidence that the dimension is in the batch table, it's just not being set in the points.

Classification is an 8bit value (uint8) in my model. And there are indeed values other than `0`:
```
      {
        "average": 7.155819695,
        "count": 21178491,
        "kurtosis": 1.176948151e+10,
        "maximum": 171,
        "minimum": 1,
        "name": "Classification",
        "position": 8,
        "skewness": 6.592046458e+10,
        "stddev": 18.91290092,
        "variance": 357.6978211
      },
```

Cesium Version 1.4.5 as part of [cesium-workshop](https://github.com/AnalyticalGraphicsInc/cesium-workshop).
Entwine Docker tag 1.3: connormanning/entwine:1.3

**Comment by devdylan** [comment]
That worked nicely! 👍 Thanks again for all your help.

**Comment by crubier** [comment]
Hi folks, awesome work, I love this !

I would love to have this particular feature in entwine:latest though :-)

**Comment by connormanning** [comment]
If someone that is using this functionality from Entwine 1.3 could make a sample list of commands, starting with a publicly accessible point cloud file, and ending with a fully functional Cesium demo that clearly verifies the batch table output, that would be of great use in getting that done, @crubier.  I don't know how people are using this functionality or how to verify it.  For example:
- Execute `docker run connormanning/entwine:1.3 build -c <some-cesium-config.json> -i <some-public-file>.laz -o output`
- Load the 3D Tiles output with `http://<some-cesium-demo-page>.html`
- If the batch table is working, `<some characteristic verifying batch table usage>` should occur.

This would provide a working sample output using the batch table which a current `entwine:master` batch table implementation could be tested against.

**Comment by crubier** [comment]
Hi @connormanning , I built a small test projet for you, so that we can reopen this issue.

Everything you need to test is there: https://github.com/crubier/EntwineTest  . Just follow the readme and you'll get started in 2 minutes.

The project is deployed on https://crubier.github.io/EntwineTest/ , which allows to see the problems there are right now: 

  - Entwine 1.3 does not correctly generate 3D tiles for cesium (some tiles fail to load)
  - Entwine latest does not generate colors and other data for cesium.

This shows we really need entwine latest to take into account additional point dimensions in cesium exports, because 1.3 does not work as it should either. Take care to not recreate the bug there is in 1.3 where buffers are not aligned properly, which causes some tiles to fail loading in cesium.

**Comment by connormanning** [comment]
Thanks for this demonstration of the issue, which is very helpful.  I will work on fixing this in the current Entwine version when I get a chance.

In the meantime you might try out [Entium](https://pypi.org/project/entium/) - a third-party project for converting EPT to 3D-Tiles written in Python.

**Comment by ProjectBarks** [comment]
@crubier I maintain that project so message me with any questions. I think it makes more sense to decouple entwines implementation of tile generation for cesium. The features you mentioned are supported with entium.

**Comment by bendidi** [comment]
hello @connormanning  and @ProjectBarks, I have the same issue as @crubier, and I've tried the solution you proposed of using `entium`, but I've run into some problems: 
 Trying to build EPT las files with `entwine` gives the following error : 
```bash
 T: 10s R: 3,578(3,578)M/h A: 12,582,912 U: 79% I: 9,940,992 P: 6% W: 0 R: 0
Exception in pool task: Invalid XYZ type
Exception in pool task: Invalid XYZ type
Exception in pool task: Invalid XYZ type
Exception in pool task: Invalid XYZ type
....
Failed GET attempt 3: /output/**/**/bin/4-3-9-11.bin
Failed GET attempt 3: /output/**/**/bin/4-0-7-14.bin
Failed GET attempt 5: /output/**/**/bin/4-2-8-12.bin
T: 30s R: 2,156(234)M/h A: 18,874,368 U: 95% I: 17,973,248 P: 11% W: 0 R: 2
Failed GET attempt 4: /output/**/**/bin/4-3-9-11.bin
Failed GET attempt 4: /output/**/**/bin/4-0-7-14.bin
......
```
The command I used is : 
``` bash
docker run -ti -v $INPUT:/input -v $OUTPUT:/output connormanning/entwine:latest build \
                    --threads $(nproc) \
                    --reprojection EPSG:4978 \
                    --dataType binary \
                    --config "/input/config.json" \
                    --input "/input/**" \
                    --output "/output/bin"
```

with config.json : 
``` json
{
  "reprojection": {
    "in": "EPSG:2154",
    "out": "EPSG:4978"
  },
  "schema": [
        { "name": "X", "type": "float" },
        { "name": "Y", "type": "float" },
        { "name": "Z", "type": "float" },
    ]
}
```
whereas when running the same but specifying the output dataType as Laszip works just fine ! , but unfortunately entium only supports `*.bin` files I believe

**Comment by connormanning** [comment]
This is being addressed via 1.3.0-maintenance in #150.

**Comment by crubier** [comment]
🎉 👍
